### PR TITLE
update linters and move their configurations from setup.cfg to pyproject.toml

### PR DESCRIPTION
**Pull request by grlee77**
The first commit here is from #615

This MR makes multiple linting-related changes to cuCIM
- moves existing `isort` config from `setup.cfg` to a new `pyproject.toml` file
- update `isort` conventions to match other RAPIDS projects
- add `black` and apply it (use `fmt::off`/`fmt::on` as needed to preserve formatting of look-up tables and test cases)
- replace use of `flake8` with `ruff`
- add config for `codespell` and apply recommended fixes across both C++ and Python code
- update CONTRIBUTING.md instructions to reflect these changes
- Pre-commit hooks for `black`, `ruff` and `codespell` were added. The pre-commit hook for `flake8` was removed.

The only non-linting change is
- move pytest configuration from `setup.cfg` to `pyproject.toml`

I plan to leave any build-system related changes to `pyproject.toml` to a follow-up MR. This MR focuses only on linting/testing.

It will probably be easiest to review the individual commits. Some, like applying `black` are quite large.

Fixes https://github.com/rapidsai/cucim/issues/23

### CODE REVIEW

**Comment by grlee77** [comment]
Most changes here are automated. 

Please closely review the following manually updated files:

- `python/cucim/setup.cfg`
- `python/cucim/pyproject.toml`
- `.pre-commit-config.yaml`

**Comment by grlee77** [comment]
Had to rename `pyproject.toml` to `pyproject_.toml` temporarily in this MR so it only gets used for linting steps, but isn't used during package build time.

**Comment by grlee77** [comment]
After this MR `black`, `ruff`, `isort` and `codespell` would all be automaticaclly run locally on each commit via `pre-commit` if the developer has previously run the following in their repository:

```
pre-commit install
``` 


By default that only runs the checks on the files being modified/added in the commit. For testing purposes, I forced pre-commit to run on all files using:
```
pre-commit run --all-files
```
to verify that no problems are reported.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -15,10 +15,23 @@ repos:
...
+  - repo: https://github.com/psf/black
+    rev: 23.3.0
     hooks:
-      - id: flake8
-        args: ["--config=python/cucim/setup.cfg"]
-        files: ^python/cucim/.*
+      - id: black
+        files: (python|legate)/.*
+        args: ["--config", "python/cucim/pyproject_.toml"]
+  - repo: https://github.com/astral-sh/ruff-pre-commit
```
Comment (bdice) [inline .pre-commit-config.yaml:25]: I won't advocate for changes in this PR due to its size, but for some potential follow-up work, you might want to update the linter versions (all RAPIDS repos' linters are independent, you don't have to use a specific isort/black/ruff version). Also you can use `ruff` in place of `isort` if you enable the `I` category, and you can enable auto-fixes as well, in `pyproject.toml`. Enabling auto-fixes plays very nicely with pre-commit.

```
[tool.ruff]
select = ["E", "F", "W", "I"]
fix = true
```

**Review by bdice — APPROVED** [review]

File: .pre-commit-config.yaml (hunk 1 above)
Comment (grlee77) [reply .pre-commit-config.yaml:25]: Sure. I also discussed with @jakirkham and @gigony that we can add a `.git-blame-ignore-revs` in a follow-up so all these linting changes [don't show up in the blame view on GitHub](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/).

**Review by raydouglass — APPROVED** [review]
`ops-codeowners` changes look good

**Review by gigony — APPROVED** [review]
Thanks @grlee77 ! Looks good to me!

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks Greg and everyone for the reviews! 🙏

File: .pre-commit-config.yaml (hunk 1 above)
Comment (jakirkham) [reply .pre-commit-config.yaml:25]: Greg opened a follow-up PR: https://github.com/rapidsai/cucim/pull/622

Please let us know if you have thoughts there 🙂

**Pull request merged**